### ISSUE TITLE
Sub label / multiline / spannable support

### ISSUE THREAD

**Issue by MFlisar** [zawadz88/MaterialPopupMenu#63]
Sometimes it may make sense to show menu entries with sub labels, can you add this as well?

E.g.:

    [SECTION]                            "Change state"
    [MENU 1]                             - "Mark completed"
    [MENU 2]                             - "Mark completed"
    [MENU 2 - SUB TEXT]                    "And add a comment"

Alternatively it would also work if we can provide a `Spannable` as text, then we can do this with it's help as well. Of course, we would also need to disable the current single line only setting for each menu entry as well.

**Comment by zawadz88** [comment]
Hi, I think this should be possible with the current version of the library (3.4.0) thanks to the recently added `viewBoundCallback` for default items as well. In `viewBoundCallback` you should be able to get the TextView and change text to a Spannable as well as and change the single line setting.

Have you tried this approach?

**Comment by MFlisar** [comment]
Did not try it yet, but it does not work because:

The super call in this block calls the callback inside `AbstractItemViewHolder`

https://github.com/zawadz88/MaterialPopupMenu/blob/ab65b75faf7b8bdebf1bc29c645e1950eeb55233/material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt#L103-L110

so whatever I do in the callback with the text will be overwritten again afterwards...

**Comment by zawadz88** [comment]
You might also try using a `customItem` then e.g.
```
customItem {
    layoutResId = R.layout.mpm_popup_menu_item
    viewBoundCallback = { view ->
        val label: TextView = view.findViewById(R.id.mpm_popup_menu_item_label)
label.apply {
            text = Html.fromHtml("<b>Some long</b> text that is applied later to see if height calculation indeed is incorrectly calculated due to this binding.")
            maxLines = 2
        }
    }
}
```
which shows like this:

![Screenshot_1561735906](https://user-images.githubusercontent.com/3322260/60353691-d9ead480-99ca-11e9-87c9-5a85d81050fb.png)

If you want to place more lines then you'd probably need to create a custom layout as well which does not have a fixed height.

**Comment by MFlisar** [comment]
If I make a pull request and add the ability to add an optional sub title instead as well, would this be ok for you? I think that this may be useful for others as well every now and then... or are you against this feature in general for some reason?

**Comment by zawadz88** [comment]
Could you post a screenshot of how would this look like?
TBH I'm not sure if this should be in the library or not and I would like to elaborate a bit why.

On one hand, as you said this might be useful to others as well. This means that the piece of code which I posted above wouldn't be needed but could be replaced with an out-of-the-box solution.

On the other hand, multiline menu items are not described in https://material.io/design/components/menus.html# so there are no guidelines on how they should be presented. I assume the solution you implemented was based on some designs, but since there are no guidelines it doesn't mean that someone else wouldn't need to restyle them anyway to e.g. change font color from black to grey or vice versa or change font size.
For all non-standard (not described in Material guideline) items I've create the `customItem` in the past so that the library would be kept simple while still allowing to do something custom e.g. add an image or multiline test labels.

**Comment by zawadz88** [comment]
I would love to hear what others think about it as well :)

**Comment by zawadz88** [comment]
On a separate note I think it would make sense for `label` to be a `CharSequence` and not a `String` to allow for `Spanned` to be set as well.

**Comment by MFlisar** [comment]
Does anything speak against moving the viewCallback to the end of the `onBind` function? This would be a good solution as well for me with nearly no drawbacks. I think this even is more natural, as a user I don't expect that changes I made inside the callback are overwritten again or that the views are not fully initialised yet (their text is not set yet)

Like e.g. here: https://github.com/zawadz88/MaterialPopupMenu/compare/master...MFlisar:master

I personally would even suggest to pass the `ViewHolder` or the views to the callback as well to avoid the heavy `findViewById` calls inside the callbacks, but this is a breaking change. Still I would consider it...

Then I can simply call following in the callback:

	   fun updateMenuText(view: View) {
			val tv = view.findViewById<TextView>(R.id.mpm_popup_menu_item_label)
			tv.setSingleLine(false)
			tv.maxLines = 2
			val text = tv.text
			val spannable = SpannableString(text)
			val idx = text.indexOf("\n")
			if (idx != -1) {
				spannable.setSpan(
					RelativeSizeSpan(0.7f),
					idx,
					text.length,
					Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
				)
				spannable.setSpan(
					ForegroundColorSpan(Tools.getThemeColor(this@MainActivity, R.attr.material_drawer_secondary_text)),
					idx,
					text.length,
					Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
				)
				tv.text = spannable
			}
		}

**Comment by zawadz88** [comment]
>Does anything speak against moving the viewCallback to the end of the onBind function? This would be a good solution as well for me with nearly no drawbacks. I think this even is more natural, as a user I don't expect that changes I made inside the callback are overwritten again or that the views are not fully initialised yet (their text is not set yet)

I like that suggestion 😀

>I personally would even suggest to pass the ViewHolder or the views to the callback as well to avoid the heavy findViewById calls inside the callbacks, but this is a breaking change. Still I would consider it...

This would probable make sense for the default items (represented by `ItemHolder`) where the layout file is known. However, this is not the case with custom items (represented by `CustomItemHolder`) and `viewBoundCallback` is declared in `AbstractItemHolder`...

It seems like the following would work as per your suggestion:
```patch
Index: sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt	(date 1561993886000)
@@ -5,6 +5,10 @@
 import android.net.Uri
 import android.os.Build
 import android.os.Bundle
+import android.text.Spannable
+import android.text.SpannableString
+import android.text.style.ForegroundColorSpan
+import android.text.style.RelativeSizeSpan
 import android.util.Log
 import android.view.Gravity
 import android.view.Menu
@@ -369,11 +373,30 @@
                         Toast.makeText(this@LightActivity, "Disabled!", Toast.LENGTH_SHORT).show()
                     }
                 }
-                customItem {
-                    layoutResId = R.layout.view_custom_item_large
+                item {
+                    label = "Line 1\nLines in secondary color"
                     viewBoundCallback = { view ->
-                        val textView: TextView = view.findViewById(R.id.customItemTextView)
-                        textView.text = "Some long text that is applied later to see if height calculation indeed is incorrectly calculated due to this binding."
+                        val tv = view.findViewById<TextView>(R.id.mpm_popup_menu_item_label)
+                        tv.setSingleLine(false)
+                        tv.maxLines = 2
+                        val text = tv.text
+                        val spannable = SpannableString(text)
+                        val idx = text.indexOf("\n")
+                        if (idx != -1) {
+                            spannable.setSpan(
+                                RelativeSizeSpan(0.7f),
+                                idx,
+                                text.length,
+                                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
+                            )
+                            spannable.setSpan(
+                                ForegroundColorSpan(ContextCompat.getColor(this@LightActivity, R.color.secondary_text_default_material_light)),
+                                idx,
+                                text.length,
+                                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
+                            )
+                            tv.text = spannable
+                        }
                     }
                 }
             }
Index: material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt	(date 1561993628000)
@@ -101,7 +101,6 @@
         private var nestedIcon: AppCompatImageView = itemView.findViewById(R.id.mpm_popup_menu_item_nested_icon)
 
         override fun bindItem(popupMenuItem: MaterialPopupMenu.AbstractPopupMenuItem) {
-            super.bindItem(popupMenuItem)
             val castedPopupMenuItem = popupMenuItem as MaterialPopupMenu.PopupMenuItem
             if (castedPopupMenuItem.label != null) {
                 label.text = castedPopupMenuItem.label
@@ -124,6 +123,7 @@
                 label.setTextColor(castedPopupMenuItem.labelColor)
             }
             nestedIcon.visibility = if (castedPopupMenuItem.hasNestedItems) View.VISIBLE else View.GONE
+            super.bindItem(popupMenuItem)
         }
     }

```
![Screenshot_1561993909](https://user-images.githubusercontent.com/3322260/60447516-d736ec00-9c23-11e9-9fce-9a9b285ee5b8.png)

What do you think about also changing `label` to `CharSequence` from `String`?
This would look like this:
```patch
Index: sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ sample/src/main/java/com/github/zawadz88/materialpopupmenu/sample/LightActivity.kt	(date 1561994499000)
@@ -5,6 +5,10 @@
 import android.net.Uri
 import android.os.Build
 import android.os.Bundle
+import android.text.Spannable
+import android.text.SpannableString
+import android.text.style.ForegroundColorSpan
+import android.text.style.RelativeSizeSpan
 import android.util.Log
 import android.view.Gravity
 import android.view.Menu
@@ -369,11 +373,12 @@
                         Toast.makeText(this@LightActivity, "Disabled!", Toast.LENGTH_SHORT).show()
                     }
                 }
-                customItem {
-                    layoutResId = R.layout.view_custom_item_large
+                item {
+                    label = createMultiLineText()
                     viewBoundCallback = { view ->
-                        val textView: TextView = view.findViewById(R.id.customItemTextView)
-                        textView.text = "Some long text that is applied later to see if height calculation indeed is incorrectly calculated due to this binding."
+                        val tv = view.findViewById<TextView>(R.id.mpm_popup_menu_item_label)
+                        tv.setSingleLine(false)
+                        tv.maxLines = 2
                     }
                 }
             }
@@ -382,6 +387,27 @@
         popupMenu.show(this@LightActivity, view)
     }
 
+    private fun createMultiLineText(): CharSequence {
+        val text = "Line 1\nLines in secondary color"
+        val spannable = SpannableString(text)
+        val idx = text.indexOf("\n")
+        if (idx != -1) {
+            spannable.setSpan(
+                RelativeSizeSpan(0.7f),
+                idx,
+                text.length,
+                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
+            )
+            spannable.setSpan(
+                ForegroundColorSpan(ContextCompat.getColor(this@LightActivity, R.color.secondary_text_default_material_light)),
+                idx,
+                text.length,
+                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
+            )
+        }
+        return spannable
+    }
+
     @OnClick(R.id.dimmedBackgroundTextView)
     fun onDimmedBackgroundClicked(view: View) {
         val popupMenu = popupMenu {
Index: material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenu.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenu.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenu.kt	(date 1561994348000)
@@ -85,12 +85,12 @@
     }
 
     internal data class PopupMenuSection(
-        val title: String?,
+        val title: CharSequence?,
         val items: List<AbstractPopupMenuItem>
     )
 
     internal data class PopupMenuItem(
-        val label: String?,
+        val label: CharSequence?,
         @StringRes val labelRes: Int,
         @ColorInt val labelColor: Int,
         @DrawableRes val icon: Int,
Index: material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/internal/PopupMenuAdapter.kt	(date 1561993628000)
@@ -101,7 +101,6 @@
         private var nestedIcon: AppCompatImageView = itemView.findViewById(R.id.mpm_popup_menu_item_nested_icon)
 
         override fun bindItem(popupMenuItem: MaterialPopupMenu.AbstractPopupMenuItem) {
-            super.bindItem(popupMenuItem)
             val castedPopupMenuItem = popupMenuItem as MaterialPopupMenu.PopupMenuItem
             if (castedPopupMenuItem.label != null) {
                 label.text = castedPopupMenuItem.label
@@ -124,6 +123,7 @@
                 label.setTextColor(castedPopupMenuItem.labelColor)
             }
             nestedIcon.visibility = if (castedPopupMenuItem.hasNestedItems) View.VISIBLE else View.GONE
+            super.bindItem(popupMenuItem)
         }
     }
 
Index: material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenuBuilder.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenuBuilder.kt	(revision ab65b75faf7b8bdebf1bc29c645e1950eeb55233)
+++ material-popup-menu/src/main/java/com/github/zawadz88/materialpopupmenu/MaterialPopupMenuBuilder.kt	(date 1561994348000)
@@ -79,7 +79,7 @@
          * Optional section holder. *null* by default.
          * If the title is not *null* it will be displayed in the menu.
          */
-        var title: String? = null
+        var title: CharSequence? = null
 
         private val itemsHolderList = arrayListOf<AbstractItemHolder>()
 
@@ -130,7 +130,7 @@
          *
          * If both [label] and [labelRes] are set [label] will be used.
          */
-        var label: String? = null
+        var label: CharSequence? = null
 
         /**
          * Item label.

```

It should allow for more flexibility without relying on custom items/ view bound callbacks for some cases. This is also a breaking change though so it should be released as next major release probably i.e. 4.0.0.

@MFlisar Please let me know what you think!

**Comment by MFlisar** [comment]
That's fine for me, using `CharSequence` makes things easier and more compact for my use case.

I still think, a callback like following would be useful:

    interface Callback<T: ViewDataBinding> {
        fun onBind(view: View, binding: T)
    }

In a large menu, calling `findViewById` a few times is noticable and not necessary, but I'm content with your suggestion as well

**Comment by zawadz88** [comment]
Cool, I'll try to have this released within the week then.

**Comment by zawadz88** [comment]
Released in 4.0.0. Thanks for reporting this!

**Comment by MFlisar** [comment]
Just checked the code and tested the new version - you moved the viewBoundCallback execution to the end of the bind item and added the `CharSequence` labels - that's perfect for me, thanks a lot
